src/schema/validator.ts

```diff
@@ -0,0 +1,185 @@
+        throw new TypeError("Feature flag enabled must be a boolean.");
+    }
+    if (featureFlag.conditions !== undefined) {
+        validateFeatureEnablementConditions(featureFlag.conditions);
+    }
+    // variants, allocation, and telemetry
+    if (featureFlag.variants !== undefined) {
+        validateVariants(featureFlag.variants);
+    }
+    if (featureFlag.allocation !== undefined) {
+        validateVariantAllocation(featureFlag.allocation);
+    }
+    if (featureFlag.telemetry !== undefined) {
+        validateTelemetryOptions(featureFlag.telemetry);
+    }
+}
+
+function validateFeatureEnablementConditions(conditions: any) {
+    if (conditions.requirement_type !== undefined && conditions.requirement_type !== "Any" && conditions.requirement_type !== "All") {
+        throw new TypeError("Feature enablement conditions requirement type must be 'Any' or 'All'.");
+    }
+    if (conditions.client_filters !== undefined) {
+        validateClientFilters(conditions.client_filters);
+    }
+}
+
+function validateClientFilters(client_filters: any) {
+    if (!Array.isArray(client_filters)) {
+        throw new TypeError("Client filter Collection must be an array.");
+    }
+
+    for (const filter of client_filters) {
+        if (typeof filter.name !== "string") {
+            throw new TypeError("Client filter name must be a string.");
+        }
+        if (filter.parameters !== undefined && typeof filter.parameters !== "object") {
+            throw new TypeError("Client filter parameters must be an object.");
+        }
+        // validate parameters here if we have schema of specific filters in the future
+    }
+}
+
+function validateVariants(variants: any) {
+    if (!Array.isArray(variants)) {
+        throw new TypeError("Variants must be an array.");
+    }
+
+    for (const variant of variants) {
+        if (typeof variant.name !== "string") {
+            throw new TypeError("Variant name must be a string.");
+        }
+        // skip configuration_value validation as it accepts any type
+        if (variant.status_override !== undefined && typeof variant.status_override !== "string") {
+            throw new TypeError("Variant status override must be a string.");
+        }
+        if (variant.status_override !== undefined && variant.status_override !== "None" && variant.status_override !== "Enabled" && variant.status_override !== "Disabled") {
+            throw new TypeError("Variant status override must be 'None', 'Enabled', or 'Disabled'.");
+        }
+    }
+}
+
+// #region Allocation
+function validateVariantAllocation(allocation: any) {
+    if (typeof allocation !== "object") {
+        throw new TypeError("Variant allocation must be an object.");
+    }
+
+    if (allocation.default_when_disabled !== undefined && typeof allocation.default_when_disabled !== "string") {
+        throw new TypeError("Variant allocation default_when_disabled must be a string.");

```

Comment:
If here we use "default_when_enabled", then all above error messages should follow this pattern.
For example, in line 44, "Feature enablement conditions requirement type must be 'Any' or 'All'." should be 
"Feature enablement conditions requirement_type must be 'Any' or 'All'."

Actually, I personally prefer to make the error message more concise like:

"'requirement_type' must be 'Any' or 'All'" But I am ok with the current error message except for the inconsistent stuff.